Expose icon SVG to theme CSS

telamonian opened the pull request:
As per @ian-r-rose's [suggestion in #5960](https://github.com/jupyterlab/jupyterlab/pull/5960#issuecomment-461880361). 

Short term, this PR will make it easier to add new icons in core; there will never be a need to make 2+ copies of each icon for light, dark, pink, etc, themes, like in the current version of `statusbar`. 

Medium term, this will enable us to automatically provide a complete set of themed icons. In in turn, this will free theme devs from having to manually repaint/manage the 100+ icons currently in a theme. 

### Details

In order to re-theme an icon using CSS, the icon's `*.svg` first has to be added to the DOM as an actual element. Our current CSS-only setup for importing SVGs isn't able to do this (I guess CSS by itself is never able to make those kinds of changes to the DOM; the closest it gets are non-stylable [replaced elements](https://developer.mozilla.org/en-US/docs/Web/CSS/Replaced_element#Replaced_elements)). So my goal is to add two things to Jupyterlab:

- A means by which to directly import `*.svg` files into `*.ts` (WIP) and `*.tsx` (done) files. This requires an `svg` specific `.d.ts` file, and some tweaks to the webpack loaders.

- A collection of objects/functions for attaching imported `svg` to the DOM. Ideally, these will be even simpler to use than the current CSS handling (which requires ~5 lines for each icon).

At the present moment, I've finished implementing all of this stuff for the `*.tsx` files in `@jupyterlab/statusbar`. I've removed the duplicate icons from this package, and have gotten all of its icons to play nicely with the different themes:

![image](https://user-images.githubusercontent.com/2263641/53341404-785e1080-38d9-11e9-8cff-2d4a0f3b030e.png)

![image](https://user-images.githubusercontent.com/2263641/53341411-7f851e80-38d9-11e9-8f9f-39fa92eb54e3.png)

telamonian commented:
I figure I'll wrap up work on this PR once I've finished exposing the remaining non-theme icons (there' 3 or 4 left).

What do people think. Does this seem beneficial?

blink1073 commented:
I think this approach looks good!

ellisonbg commented:
Fantastic, thanks for working on this, it will really help us. A few questions...

* Right now most of our icons are in the themes. Are you thinking of collecting the icons into a separate new or existing jlab package so lots of other packages can use them?
* Our current icon set is a mix of Material Design icons, and a few custom ones. Because we have had to manually change the colors of the svg's, we have been manually copying the MD icons into the themes. I would love to migrate over the depending on a well maintained npm package for the MD icons. Do you think this is possible?
* We are beginning to use blueprint for react components. Have you looked at how they handle icons to make sure this approach will work with this one?

Sorry for all the questions - excited about this direction.

ian-r-rose commented:
@telamonian what is the status of this? It would be great to get it in :)

gnestor commented:
@telamonian We are getting very close to a 1.0 release and I agree that this would be good to get in there! Can I help push this past the finish line?

telamonian commented:
I took some time to finish defending my thesis, and then I was dead to the world for a while, but I'm back on this now.

@ellisonbg I have answers to your questions now:

- I do think it would be best to collect all of the icons into a single core package. Right now to me `ui-components` seems like the best place for them.

- If you can point me towards such an npm package that has a collection of the MD icons, I can let you know if it's feasible to use. Basically, I think it'll come down to how the package in question exposes it's `.svg` files.

- I looked into the blueprint icon stuff. My approach won't interfere with the icons that blueprint provides. I think it may in fact be possible to unify my icon stuff with the existing infrastructure in blueprint. However, I ran into some gnarly build issues when I tried this, so I think that'll have to wait for a future PR.

ian-r-rose commented:
Congratulations on defending, Max!

gnestor commented:
Congrats @telamonian!

To respond to your questions:

* Agreed, I would definitely like to see the icons live here. We only need to keep the custom icons here and then we can use a Material Design icon font or npm package that contains all of the SVGs.

* You can import the official Material Design icon font

```html
<link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
```

If you need the SVG files, I don't immediately see an npm package that provides them, but they're all in https://github.com/google/material-design-icons. 

* I'm happy to help with the blueprint-related stuff. Blueprint provides a pretty good icon set so we could consider using it.

telamonian commented:
Apologies for the delay. Partly I've been busy (now that I'm no longer a student, I need to find paying work), and partly I ran head first into a bunch of issues with Phosphor, and how it handles icons.

Good news, though @gnestor! I fished your wish, and now the tab icons on the sidebar are all themable. Extensions devs will have to set up their icons via my `IconRegistry` stuff (and I'll have to do some more organization in order to make that process simple), but in the current version of this PR most of the tab icon stuff has been automated.

ian-r-rose commented:
@telamonian I know some folks at Berkeley who may be interested in hiring a scientist with Jupyter experience 😉

telamonian commented:
There are currently two roadblocks to finishing this PR, one simple and one probably quite difficult:

- I'm having trouble finding a set of colors to use with the icons that a) already exist in the themes and b) reproduce the existing icon colors. All of my icon color CSS is in `packages/application/style/images.css`, and I added relevant CSS classes directly to the `.svg` files in `packages/ui-components/style`.

- Phosphor handles the creation/rendering of many of jlab's tab icons (eg in the sidebar tabs and topbar document tabs), and does so in an opaque fashion. By overriding some of Phosphor's default VDOM renderers, I've made some headway in getting `.svg` icons in the tabs. However, my [current solutions](https://github.com/jupyterlab/jupyterlab/pull/6034/files#diff-62da0550d0f6d9bceb4bb477732a8e38R1201) all rely on [a hack](https://github.com/jupyterlab/jupyterlab/pull/6034/files#diff-28e59bd085d6bc0a9f9168f021bb1ca3R145) whereby the icon nodes are retrieved after their initial creation via `document.getElementsByClassName`. This is not ideal for a bunch of reasons (including performance). The way I see it, there are two possible development paths:
    1. Figure out how to smoothly integrate `IconRegistry` with Phosphor. I think this can be done by passing custom VDOM renderers to certain widgets via their constructors. However, I suspect a complete version of this solution will require an upgrade of [@phosphor/virtualdom](https://github.com/phosphorjs/phosphor/blob/e4ef73a881d902ca00d9e991a228282a977c3fb8/packages/virtualdom/src/index.ts#L27) in order to work.
    2. Work on polishing up my current hacks to ensure they're robust and do the minimum required work.

I'll keep working on it. @gnestor @ian-r-rose (or anyone) If I could get a hand working out these last few issues it would be greatly appreciated.

Also, Ian, that sounds interesting. Possibly very interesting. I'll email you to follow up.

telamonian commented:
I figured out some reasonable ways to deal with the phosphor issues (via some light subclassing). Functionally, I think this PR is done. 

Here's what the PR covers:
- Tools for adding inline svg icons, mostly encapsulated in the `IconRegistry` class. All of this stuff has been designed to play nicely with the existing icon handling (as css background images), so both can co-exist.

- All existing `.svg` files in core (including all of the left sidebar tab icons, excluding the hundred-or-so remaining icons in the theme packages) have been moved to `ui-components`, and are being handled by `IconRegistry`. I've added classes (and associated CSS) to the elements of `.svg` files so they repaint correctly in accordance with the theme.

There are still a number of issues, but they largely all have to do with integrating the inline svg icon stuff better with core and extensions.

Right now seems like a good time to get some further review/feedback. @saulshanabrook @blink1073 @ellisonbg @ian-r-rose @gnestor any input you can give me would be greatly appreciated.

tslaton commented:
I was looking into adding icons matching the style of JupyterLab core to an extension I'm working on, and upon struggling to style those icons, I came across this issue. I think the progress sounds good!

Since this is a large rework of all the icons anyway, I wonder: are they licensed correctly currently?

Google says their Material Design icons are licensed under Apache 2.0, but I don't see the license included anywhere. I was wondering how to include it myself. 

Curiously on their own copy of the license, they appear to have failed to register their copyright (https://github.com/google/material-design-icons/issues/912).

telamonian commented:
@tslaton I can honestly say that I have no idea whatsoever. But if you want to dig into it I'd be glad to give you a hand. I'll probably save it for the next PR, though. This one has already creeped all of the scopes.

telamonian commented:
I did some profiling using devtools. I ran the refresh task 10 times for both this PR and master, and recorded the times:

- master: `3416 +/- 206 ms`
- this PR: `3171 +/- 210 ms`

So if anything, the icon handling in this PR (inline `svg`) is actually slightly faster than the existing icon handling (CSS background images). I'm not sure if this trend holds across all of the possible icon rendering tasks, but it at least seems safe to say that this PR does not have a negative impact on jlab's overall performance.

telamonian commented:
I think this PR is ready for a final review and wrap-up (hopefully before I have to rebase it an eighth time). There's a handful of small issues:

- The home icon above the filebrowser has gone missing
- The icons in the settings editor have gone a bit weird
- A number of icons have changed since I started working on this PR and need to be updated

I'm already working on these. There's also a number of issues that I think are best saved for a future PR (this one has sprawled enough):
- Finish inlining all svg icons (ie moving all of the icons out of `theme-x-extension` and into `ui-components`)
- Finish refactoring the `defaultIconProvider` singleton into a replaceable plugin

Pinging @ellisonbg @saulshanabrook for a review.

telamonian commented:
@saulshanabrook Those are some nice and detailed notes. One thing to add:

I really like the compile-time correctness of your `export const runningIcon = new Icon(runningSVG)` idea. But. After we got off the call though I realized that we'll probably still need an `IconRegistry` to deal with the cases in which the icon is set dynamically based on a variable.

For example, in `docregistry` the tab bar icon of a document is determined via the `fileType` of the `MimeContent` widget, and is then passed to that widget's `title`:

https://github.com/jupyterlab/jupyterlab/blob/bc65c2743f46a41cc6c095edaa57b65f588576c7/packages/docregistry/src/mimedocument.ts#L288-L289

In a custom renderer I then pick the correct svg based on `title.iconClass`:

https://github.com/jupyterlab/jupyterlab/blob/aa1e96ee41533c698fd57573c08aba15350e909f/packages/ui-components/src/icon/tabbarsvg.ts#L41-L48

If each icon is it's own importable object, how would you pick the relevant one here?

On the filthy hack side of things, I suppose you could surmount the issue by widening `title` at the points of consumption:

```typescript
class wideTitle extends Title<Widget> {
  iconSvg: Icon;
}

...

let title: wideTitle = content.title;

title.iconSvg = ft.iconSvg;
```

Probably not the best solution. Better would be to go upstream to Phosphor and push some kind of support for `title.iconSvg` (we could maybe even repurpose the deprecated `title.icon` parameter). Until that's settled I think we're stuck with `IconRegistry`.

saulshanabrook commented:
> Better would be to go upstream to Phosphor and push some kind of support for title.iconSvg (we could maybe even repurpose the deprecated title.icon parameter).

Yeah that would make sense to me!

telamonian commented:
@saulshanabrook The PR is done* and ready for testing. The PR should now have complete parity (minus a couple of minor style changes that I couldn't avoid) with master, though I would not be surprised if I missed something. Lemme know how it goes.

**There's one more styling task (for the icons in the tab manager in the sidebar), and I need to figure out why some tests related to the home breadcrumb are failing, but those shouldn't stop you*

telamonian commented:
> though I would not be surprised if I missed something

and I've already found the first thing I missed: the icons in the launcher are screwed up.

ellisonbg commented:
Important question: is this API breaking?

On Wed, Jul 10, 2019 at 9:52 PM Max Klein <notifications@github.com> wrote:

> though I would not be surprised if I missed something
>
> and I've already found the first thing I missed: the icons in the launcher
> are screwed up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/jupyterlab/pull/6034?email_source=notifications&email_token=AAAGXUD73BQTLAQAY2M2C23P6237PA5CNFSM4GZ7YQ6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZVP32A#issuecomment-510328296>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGXUGX5VKM34YWR44M5ALP6237PANCNFSM4GZ7YQ6A>
> .
>


-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

saulshanabrook commented:
@ellisonbg I believe yes, since any extensions that depend on the current CSS classes for these icons will stop working. Is that right max?

telamonian commented:
@ellisonbg No functionality will break. If an extension is reusing one of the icons that I've moved into `ui-components` via its current CSS class, that icon will appear blank.

ellisonbg commented:
The functionality is not the part I was asking about (I was assuming it is preserved). Our CSS classes are part of our public API so we should not change them in a way that will break extensions until the next major release. Right now the CSS API is a mess and dependencies and versions are not clear - we have plans to improve that but need to not break things in the meantime.

Sent from my iPhone

> On Jul 11, 2019, at 8:50 PM, Max Klein <notifications@github.com> wrote:
> 
> @ellisonbg No functionality will break. If an extension is reusing one of the icons that I've moved into ui-components via its current CSS class, that icon will appear blank.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

telamonian commented:
@ellisonbg If it's an issue, for now I can just re-add the icon CSS classes that have been removed. I do think that the icon CSS classes should removed so that:

- there's exactly one correct way (that plays nicely with the theming system) to consume icons in Jlab
- the theming issues (pointed out [earlier in this thread](https://github.com/jupyterlab/jupyterlab/pull/6034#issuecomment-484701867) by @gnestor) for icons supplied by extensions get fixed

but as you say, that can wait for the next major version.

jasongrout commented:
We did just release a few days ago, and most extension developers probably haven't ported things to 1.0 yet. I think it's probably fine to deprecate those things (already?!) and have one opinionated way to do things.

telamonian commented:
I fixed the issues with the launcher icons. Anyone who tests this, let me know if I missed anything else.

telamonian commented:
> We currently import all the icons with [a glob webpack `require.context`](https://github.com/jupyterlab/jupyterlab/pull/6034/files#diff-e69bc6c8786da9c3830d3849aecaed7cR86) to get all the icons in a certain directory:

I was thinking through how I'd implement `test-ui-components`, and I realized that as long as I'm using the `require.context` glob import (which doesn't work in Jest) then there's no good way.

I replaced the `require.context` with explicit `import` statements, and added an `ensureUiComponents` function to `buildutils` that keeps the `import`s synced with the contents of `ui-components/style/icons`.

saulshanabrook commented:
> I replaced the require.context with explicit import statements, and added an ensureUiComponents function to buildutils that keeps the imports synced with the contents of ui-components/style/icons.

Nice!

saulshanabrook commented:
> I think it's probably fine to deprecate those things (already?!) and have one opinionated way to do things.

So @jasongrout would you recommend preserving backwards compatibility then? I heard a rumor we weren't having any breaking changes for six months...

jasongrout commented:
We need to preserve backwards compatibility at this point, I think, but it's fine to me if we deprecate things

ellisonbg commented:
At what point will we create a 1.x branch and start to break things in
master for 2.0?

On Mon, Jul 15, 2019 at 1:07 PM Jason Grout <notifications@github.com>
wrote:

> We need to preserve backwards compatibility at this point, I think, but
> it's fine to me if we deprecate things
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/jupyterlab/pull/6034?email_source=notifications&email_token=AAAGXUAI62YRURCJCXKW5CDP7TKJPA5CNFSM4GZ7YQ6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ62DTQ#issuecomment-511549902>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGXUALSUIAZSAU2JUJ7DTP7TKJPANCNFSM4GZ7YQ6A>
> .
>


-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

telamonian commented:
@ellisonbg @jasongrout For the purpose of this PR, what's the consensus? Here's my sense of what's being asked for:

- re-add/preserve the existing icon CSS classes
- mark the existing icon CSS classes as deprecated (at least in the comments)

Does that sound reasonable to both of you?

telamonian commented:
> Important question: is this API breaking?

@ellisonbg Not in the latest revision of this PR, no. I de-duped a whole bunch of CSS and icon files, and ensured that all of the old CSS classes have been preserved and added to an appropriate file in `ui-components`.

> I think it's probably fine to deprecate those things (already?!) and have one opinionated way to do things.

@jasongrout For all of the icons that have been moved to `ui-components` already, I moved the declarations of their CSS classes to `ui-components/style/deprecated.css`. There's also a set of icons (the "selected" versions of certain icons) that are no longer needed, which I moved to `ui-components/deprecated/selected`. Is there anything else I should do to mark the CSS as deprecated?

vidartf commented on buildutils/src/ensure-package.ts:
```diff
@@ -138,17 +168,20 @@ export async function ensurePackage(
+
+    // write out cssIndexContents, if needed
+    const cssIndexPath = path.join(pkgPath, 'style/index.css');
+    messages.push(...ensureFile(cssIndexPath, cssIndexContents));
```
I think the previous code would create the file if it didn't exist, but this function will fail?

telamonian replied:
If the file doesn't exist, the old code will raise an error when it tries to read at:

```typescript
const prev = fs.readFileSync(cssPath, { encoding: 'utf8' });
```

The new code will instead print a warning if the file doesn't exist. Not creating the file makes sense, since this would limit possible weird consequences if the `buildutils` is out of sync with the rest of the packages, for whatever reason.

telamonian replied:
If the file doesn't exist, the old code would raise an error when it tried to read:

```typescript
const prev = fs.readFileSync(cssPath, { encoding: 'utf8' });
```

The new code will print a warning. Not creating the file makes sense, since it minimizes possible consequences if the version of `buildutils` is out of sync with the rest of the packages.

vidartf replied:
> If the file doesn't exist, the old code would raise an error when it tried to read

Ok, I agree that is good to clean up. Another issue though: Does not the packagess with only an `index.css` (and no `base.css`) need this treatment? Or is the way SVGs are consumed intrinsically link them to the base/index split setup?

telamonian replied:
This section of the code has nothing to do with the new svg handling. I needed to copy some functionality from this section and use it several times in `ensureUiComponents`, so I encapsulated that functionality in a new `ensureFile` function. I then cleaned up the old code by replacing it with `ensureFile`, which also has the benefit of giving it the more graceful error handling.

> Or does the way SVGs are consumed intrinsically link them to the base/index split setup?

I don't think there's any cases in which a split `index.css` and `base.css` are actually required. We just split them up that way so that one part (`index.css`) can be 100% automatically managed by `buildutils`.

telamonian commented:
And another rebase finished!

1.1dev is coming soon, right?

blink1073 commented:
@jasongrout, are you available to meet sometime in the next couple of days to push this forward?

jasongrout commented:
Sure. But I don't think we have to wait to merge this, if it's ready to go in.

blink1073 commented:
Thanks for your hard work and patience on this, @telamonian!

ian-r-rose commented:
Thanks for the herculean effort @telamonian!

telamonian commented:
I'm just very happy right now that I will never have to rebase this PR ever again. Thanks guys.
